zebedee-reader/src/main/java/com/github/onsdigital/zebedee/util/URIUtils.java

```diff
@@ -79,11 +83,93 @@ public static String getLastSegment(String uri) {
         return segments[segments.length - 1];
     }
 
-    public static String removeLastSegment(String uri) {
-        if (isEmpty(uri)) {
-            return null;
+    public static String removeAfterKeyword(String path, String keyword) {
+        int index = path.indexOf(keyword);
+        if (index == -1) {
+            return path;
+        }
+        String result = path.substring(0, index);
+        return result.endsWith("/") ? result.substring(0, result.length() - 1) : result;
+    }
+
+    public static String getQueryParameterFromURL(String url, String queryParam) {
+        try {
+            URI uri = new URI(StringUtils.lowerCase(url));
+            String query = uri.getQuery();
+            if (query != null && query.contains(StringUtils.lowerCase(queryParam))) {
+                Map<String, String> queryParameterMap = getParameterMapFromQuery(query);
+                String paramValue = queryParameterMap.get(StringUtils.lowerCase(queryParam));
+                return getDecodedUrl(paramValue);
+            }
+        } catch (URISyntaxException e) {
+            throw new RuntimeException("Error retrieving query parameter from path", e);
+        }
+        return null;
+    }
+
+    public static String getDecodedUrl(String url) {
+        if (isUUEncoded(url)) {
+            return URLDecoder.decode(url);
+        }
+        return url;
+    }
+
+    public static boolean isUUEncoded(String url) {
+        return url.toLowerCase().contains("%2f");
+    }
+
+    public static boolean isLastSegmentFileExtension(String path) {
+        if ("".equals(path)){
+            return false;
+        }
+        String lastSegment = getLastSegment(path);
+        if (lastSegment == null){
+            return false;
         }
-        String result =  removeEnd(removeTrailingSlash(uri), FORWARD_SLASH + getLastSegment(uri));
+        return lastSegment.contains(".");
+    }
+
+    public static boolean hasFileExtension(String path) {
+        return path.contains(".");
+    }
+
+    public static String removeLastSegment(String uri) {
+        String result = removeEnd(removeTrailingSlash(uri), FORWARD_SLASH + getLastSegment(uri));
         return isEmpty(result) ? "/" : result;
     }
+
+    public static String getNSegmentsAfterSegmentInput(String uri, String segmentInput, int n) {
+        if (!uri.contains(segmentInput)) {
+            return uri;
+        }
+        int start = uri.indexOf(segmentInput);
+        String everythingAfterSegment = uri.substring(start + segmentInput.length());
+        return uri.substring(0, start + segmentInput.length()) + getNSegments(removeLeadingSlash(everythingAfterSegment), n);
+    }
+
+    public static String getNSegments(String uri, int n) {
+        String result = "";
+        String parts[] = removeLeadingSlash(uri).split("/");
+        for (int i = 0; i < Math.min(parts.length, n); i++) {
+            if (parts[i] != null) {
+                result = result + "/" + parts[i];
```

Comment:
You could consider doing this with a `StringBuilder`, as it would be more efficient.